src/best-practices/caddy.rst

```diff
@@ -0,0 +1,167 @@
+requests from ``http(s)://domain.com/...`` to ``http://localhost:5984/...``
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy / localhost:5984 {
+            transparent
+        }
+
+    }
+
+Note that, because Caddy is https-by-default, you must explicitly include the
+``http://`` protocol in the site address if you do NOT want Caddy
+to automatically acquire and install an SSL certificate and begin accepting
+``https`` connections on port 443.
+
+Reverse proxying CouchDB in a subdirectory with Caddy
+=====================================================
+
+It can be useful to provide CouchDB as a subdirectory of your overall domain,
+especially to avoid CORS concerns. Here's an excerpt of a basic Caddy
+configuration that proxies the URL ``http(s)://domain.com/couchdb`` to
+``http://localhost:5984`` so that requests appended to the subdirectory, such
+as ``http(s)://domain.com/couchdb/db1/doc1`` are proxied to
+``http://localhost:5984/db1/doc1``.
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy /couchdb localhost:5984 {
+            transparent
+            without /couchdb
+        }
+
+    }
+
+Note that in the above configuration, the *Verify Installation* link in
+Fauxton may not succeed.
+
+Reverse proxying + load balancing for CouchDB clusters
+======================================================
+
+Here's a basic excerpt from a Caddyfile in
+``/<path>/<to>/<site>/Caddyfile``. This will proxy and evenly distribute all
+requests from ``http(s)://domain.com/...`` among 3 CouchDB cluster nodes
+at ``localhost:15984``, ``localhost:25984`` and ``localhost:35984``.
+
+Caddy will check the status, i.e. health, of each node every 5 seconds;
+if a node goes down, Caddy will avoid proxying requests to that node until it
+comes back online.
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy / http://localhost:15984 http://localhost:25984 http://localhost:35984 {
+            policy round_robin
+            health_check /_up
+            health_check_duration 5s
+            try_interval 500ms
+            timeout 1h
+            transparent
+        }
+
+    }
+
+Note that in the above configuration, the *Verify Installation* link in
+Fauxton may not succeed.
+
+Authentication with Caddy as a reverse proxy
+============================================
+
+Here's a sample config setting with basic authentication enabled, placing
+CouchDB in the ``/couchdb`` subdirectory:
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        basicauth /couchdb couch_username couchdb_password
+
+        proxy /couchdb localhost:5984 {
+            transparent
+            header_upstream -Authorization
```

Comment:
Is an example available that correctly passes the `X-Auth-CouchDB-*` headers downstream? I didn't write one for Nginx, but arguably there should be one in there at some point. See http://docs.couchdb.org/en/stable/api/server/authn.html#proxy-authentication for details.

Reply:
Caddy does not add, remove or alter up- or downstream headers unless explicitly instructed to do so. So in this case it didn't seem necessary as it's default behavior.

Reply:
> If you specifically want to "login" to CouchDB as a user when set up in proxy auth mode, you specify `X-Auth-CouchDB-UserName: wohali` and possibly also e.g. `X-Auth-CouchDB-Roles: admin,writer`.

I got the impression somewhere that `proxy_authentication_handler` is on its way out... is that not the case?

Reply:
No, CouchDB proxy (where CouchDB is the reverse proxy for another service) is gone. CouchDB auth proxy will stay for the foreseeable future...